firefox-ios/Client/Frontend/Home/Homepage Rebuild/TopSites/TopSitesManager.swift

```diff
@@ -205,4 +219,30 @@ class TopSitesManager: TopSitesManagerInterface, FeatureFlaggable {
             return state
         }
     }
+
+    // MARK: - Context menu actions
+    func removeTopSite(_ site: Site) {
+        unpinTopSite(site)
+        dispatchQueue.asyncAfter(deadline: .now() + 0.3) { [weak self] in
+            self?.hideURLFromTopSites(site)
+        }
+    }
+
+    func pinTopSite(_ site: Site) {
+        _ = profile.pinnedSites.addPinnedTopSite(site)
+    }
+
+    func unpinTopSite(_ site: Site) {
+        googleTopSiteManager.removeGoogleTopSite(site: site)
+        topSiteHistoryManager.removeTopSite(site: site)
+    }
+
+    private func hideURLFromTopSites(_ site: Site) {
+        topSiteHistoryManager.removeDefaultTopSitesTile(site: site)
+        // We make sure to remove all history for URL so it doesn't show anymore in the
+        // top sites, this is the approach that Android takes too.
+        profile.places.deleteVisitsFor(url: site.url).uponQueue(.main) { [weak self] _ in
+            NotificationCenter.default.post(name: .TopSitesUpdated, object: self)
+        }
```

Comment:
mainly copied these methods over from existing implement, only change was adding `weak self` since the tests caught a memory leak